deno_dist/request.ts

```diff
@@ -117,52 +77,129 @@ export function extendRequestPrototype() {
       }
       return result
     }
-  } as InstanceType<typeof Request>['queries']
+  }
+
+  header(name: string): string
+  header(): Record<string, string>
+  header(name?: string) {
+    if (!this.headerData) {
+      this.headerData = {}
+      this.original.headers.forEach((value, key) => {
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+        this.headerData![key] = value
+      })
+    }
+    if (name) {
+      return this.headerData[name.toLowerCase()]
+    } else {
+      return this.headerData
+    }
+  }
 
-  Request.prototype.cookie = function (this: Request, key?: string) {
-    const cookie = this.headers.get('Cookie') || ''
+  cookie(key: string): string | undefined
+  cookie(): Cookie
+  cookie(key?: string) {
+    const cookie = this.original.headers.get('Cookie')
+    if (!cookie) return
     const obj = parse(cookie)
     if (key) {
       const value = obj[key]
       return value
     } else {
       return obj
     }
-  } as InstanceType<typeof Request>['cookie']
+  }
 
-  Request.prototype.parseBody = async function <BodyType extends BodyData>(
-    this: Request
-  ): Promise<BodyType> {
+  async parseBody<BodyType extends BodyData>(): Promise<BodyType> {
     // Cache the parsed body
     let body: BodyType
     if (!this.bodyData) {
-      body = await parseBody<BodyType>(this)
+      body = await parseBody<BodyType>(this.original)
       this.bodyData = body
     } else {
       body = this.bodyData as BodyType
     }
     return body
-  } as InstanceType<typeof Request>['parseBody']
+  }
 
-  Request.prototype.json = async function <JSONData = unknown>(this: Request) {
+  async json<JSONData = unknown>() {
     // Cache the JSON body
     let jsonData: Partial<JSONData>
     if (!this.jsonData) {
-      jsonData = JSON.parse(await this.text())
+      jsonData = JSON.parse(await this.original.text())
       this.jsonData = jsonData
     } else {
       jsonData = this.jsonData
     }
     return jsonData
-  } as InstanceType<typeof Request>['jsonData']
+  }
+
+  async text() {
+    return await this.original.text()
+  }
+
+  async arrayBuffer() {
+    return await this.original.arrayBuffer()
+  }
+
+  async blob() {
+    return await this.original.blob()
+  }
+
+  async formData() {
+    return await this.original.formData()
+  }
```

Comment:
I would avoid the `await`s here to prevent unnecessary microtasks.

Performance impact is probably negligible because engines can optimize for these scenarios (but still, maybe not all of them), but it also changes the execution flow in subtle ways.

See: https://v8.dev/blog/fast-async

Reply:
> I would avoid the awaits here to prevent unnecessary microtasks.

This is good. And we can also avoid the await inside `c.json()`.